File: server/datasets/python_source-code/Multiple-G2/structural_30_avl_tree.py
class Node_5:
    def __init__(self, v_5):
        self.v_5 = v_5
        self.r_5 = None
        self.l_5 = None
        self.height = 1

def h_5(node):
    if not node:
        return 0
    return node.height

def b_5(node):
    if not node:
        return 0
    return h_5(node.l_5) - h_5(node.r_5)

def rr_5(y):
    x = y.l_5
    sub = x.r_5
    x.r_5 = y
    y.l_5 = sub
    y.height = 1 + max(h_5(y.l_5), h_5(y.r_5))
    x.height = 1 + max(h_5(x.l_5), h_5(x.r_5))
    return x

def rl_5(x):
    y = x.r_5
    sub = y.l_5
    y.l_5 = x
    x.r_5 = sub
    x.height = 1 + max(h_5(x.l_5), h_5(x.r_5))
    y.height = 1 + max(h_5(y.l_5), h_5(y.r_5))
    return y

def ins_5(rt_5, v_5):
    if not rt_5:
        return Node_5(v_5)
    if v_5 < rt_5.v_5:
        rt_5.l_5 = ins_5(rt_5.l_5, v_5)
    elif v_5 > rt_5.v_5:
        rt_5.r_5 = ins_5(rt_5.r_5, v_5)
    else:
        return rt_5

    rt_5.height = 1 + max(h_5(rt_5.l_5), h_5(rt_5.r_5))
    bf_5 = b_5(rt_5)

    if bf_5 > 1 and v_5 < rt_5.l_5.v_5:
        return rr_5(rt_5)
    if bf_5 < -1 and v_5 > rt_5.r_5.v_5:
        return rl_5(rt_5)
    if bf_5 > 1 and v_5 > rt_5.l_5.v_5:
        rt_5.l_5 = rl_5(rt_5.l_5)
        return rr_5(rt_5)
    if bf_5 < -1 and v_5 < rt_5.r_5.v_5:
        rt_5.r_5 = rr_5(rt_5.r_5)
        return rl_5(rt_5)

    return rt_5

File: server/datasets/java_source-code/Multiple-G2/structural_30_avl_tree.java
class Node_5 {
    int k_5, h_5;
    Node_5 left, right;
    Node_5(int d) {
        k_5 = d;
        h_5 = 1;
    }
}
public class Tree_5 {
    public static void main(String[] args) {
        Tree_5 t = new Tree_5();
        Node_5 r = null;
        r = t.ins_5(r, 10);
    }
    int max(int a, int b) {
        return (a > b) ? a : b;
    }
    int h_5(Node_5 N) {
        if (N == null) return 0;
        return N.h_5;
    }
    int gb_5(Node_5 N) {
        if (N == null) return 0;
        return h_5(N.left) - h_5(N.right);
    }
    Node_5 upd_5(Node_5 n_5) {
        n_5.h_5 = 1 + max(h_5(n_5.left), h_5(n_5.right));
        return n_5;
    }
    Node_5 rr_5(Node_5 y) {
        Node_5 x = y.left;
        Node_5 T2 = x.right;
        x.right = y;
        y.left = T2;
        upd_5(y);
        upd_5(x);
        return x;
    }
    Node_5 rl_5(Node_5 x) {
        Node_5 y = x.right;
        Node_5 T2 = y.left;
        y.left = x;
        x.right = T2;
        upd_5(x);
        upd_5(y);
        return y;
    }
    Node_5 ins_5(Node_5 n_5, int k_5) {
        if (n_5 == null) return (new Node_5(k_5));
        if (k_5 < n_5.k_5) n_5.left = ins_5(n_5.left, k_5);
        else if (k_5 > n_5.k_5) n_5.right = ins_5(n_5.right, k_5);
        else return n_5;
        upd_5(n_5);
        int b_5 = gb_5(n_5);
        if (b_5 > 1 && k_5 < n_5.left.k_5) return rr_5(n_5);
        if (b_5 < -1 && k_5 > n_5.right.k_5) return rl_5(n_5);
        if (b_5 > 1 && k_5 > n_5.left.k_5) {
            n_5.left = rl_5(n_5.left);
            return rr_5(n_5);
        }
        if (b_5 < -1 && k_5 < n_5.right.k_5) {
            n_5.right = rr_5(n_5.right);
            return rl_5(n_5);
        }
        return n_5;
    }
}